### DISCUSSION TITLE
Support for BLE pheripheral on CH32V208

### DISCUSSION THREAD

**Discussion by j4ger** [ch32-rs/ch32-hal#30] [General]
Hi all, first of all, I'd like to thank all of you for this amazing project, it's nice to see rust-embedded ecosystem growing.
Currently I'm planning on making a bluetooth keyboard and looking for a mcu with ble and usb-peripheral capabilities. CH32V208 and CH58x series meet my needs perfectly so I further researched on their rust support situation. I was glad to find out that there has been a rather nicely-done hal for CH32 series, but only BLE peripheral support is missing. I'm wondering if this can be implemented in the future, as it is not listed in the `TODOs` section.
Btw I noticed that CH58x-hal has implemented BLE functions (but no USB-related drivers), is it possible to port the logic here? I'm not very familiar with the underlying core differences of the two series so I may be asking stupid questions.

**Comment by andelf** [comment]
WCH provides BLE as closed-source libraries (LibWCHBLE.a in the EVT codebase). It includes a simple BLE programming manual PDF. CH58x-hal is an experiment for integrating BLE and rust-embedded, and the result seems promising:

- WCH's BLE library uses a lightweight RTOS (TMOS) to handle the BLE stack.
- TMOS has some extension points (task callback, task events), so if the time slicing is handled carefully, it can work with the Embassy framework.
- All WCH's BLE MCU provide almost the same BLE API. So most BLE glue code can be reuseable.

Yes, the USB part is still missing. It might require 1 or 2 weeks of work to finish it.

**Reply by j4ger** [reply]
I see. Glad to hear that things are working out! I still have two more questions though:

- Other chip functionalities aside, is it better to use CH32 series than CH58x/59x series chips if I want the most mature rust support? I guess CH58x-hal should be regarded as an experiment and is subject to change in the future?
- From the description of the Qingke cores (based on wch.cn), it seems like both CH59x and CH32V208 contains the same V4C core, is it possible to unify their hal(s)? Or, am I having some misunderstanding about their differences, it could be unrelated with the inner core I guess?

**Reply by andelf** [reply]
> it seems like both CH59x and CH32V208 contains the same V4C core

Yes, the RISC-V cores are the same, but the peripherals are not. CH57x, CH58x, CH59x, and CH56x (USB 3.0 SuperSpeed ones) use the same set of basic peripherals: GPIO and RTC are not split as separate blocks; they are mixed with the SYS block. So they are entirely different MCU families.


> is it better to use CH32 series than CH58x/59x series chips

The CH32V208 has more RAM and a higher clock speed. WCH officially states that the CH32V208 can be treated as a CH32V203 with the CH582's BLE (according to a CTO's tweet).

I'd say both are OK for new design choices. The CH58x-hal might be migrated to a unified CH5xx-hal in the future, but it's not planned yet.

If you are planning to use Rust for your project, the CH58x-hal's BLE FFI binding is a good reference, despite the missing USB part.

**Reply by j4ger** [reply]
I understand now, thanks for the explanation! I'll be keeping my eyes on the project then. Really looking forward to the unified implementation, as these chips can be a lot less power-hungry than ESP32 series and all they need is proper software.

**Reply by jsprog** [reply]
@j4ger & @andelf 

Hardware agnostic Host already exists `trouble_host` and I'd rather suggest adding the external controller by implementing `bt_hci::controller::ExternalController` for CH32V208 and CH5XX series

[https://embassy.dev/trouble/](https://embassy.dev/trouble/)
[https://github.com/embassy-rs/trouble](https://github.com/embassy-rs/trouble)
[https://github.com/embassy-rs/trouble/tree/main/examples](https://github.com/embassy-rs/trouble/tree/main/examples)
[https://github.com/embassy-rs/bt-hci](https://github.com/embassy-rs/bt-hci)
